app/app.go

```diff
@@ -0,0 +1,105 @@
+package app
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/app"
+	"github.com/iov-one/weave/coin"
+	"github.com/iov-one/weave/orm"
+	"github.com/iov-one/weave/store/iavl"
+	"github.com/iov-one/weave/x"
+	"github.com/iov-one/weave/x/msgfee"
+	"github.com/iov-one/weave/x/multisig"
+	"github.com/iov-one/weave/x/sigs"
+	"github.com/iov-one/weave/x/utils"
+)
+
+// Authenticator returns authentication with multisigs
+// and public key signatues
+func Authenticator() x.Authenticator {
+	return x.ChainAuth(sigs.Authenticate{}, multisig.Authenticate{})
+}
+
+// Chain returns a chain of decorators, to handle authentication,
+// fees, logging, and recovery
+func Chain(authFn x.Authenticator, minFee coin.Coin) app.Decorators {
+
+	// TODO implement orderbook controller
+	return app.ChainDecorators(
+		utils.NewLogging(),
+		utils.NewRecovery(),
+		utils.NewKeyTagger(),
+		utils.NewSavepoint().OnCheck(),
```

Comment:
Can you explain to me the use cases of 2 Savepoints: `OnCheck` and `OnDeliver`? Which one should I pick depending on what?

Reply:
Basically what savepoint does is wraps datastore in a cache before passing it to further handlers.
So if `Check` in a handler fails - it won't affect the state even if it wrote something to the cache.

If it does not fail, however, the cache is being dumped to the db.
Deliver check does the same for deliver.

Reply:
Not that bnsd does not use `OnDeliver` savepoint, because this is implemented in `dynamicFee` extension and applied in bnsd.